### PR TITLE
12 2.1 change application until deadline

### PR DESCRIPTION
**Pull request by nstrelow**

### CODE REVIEW

File: app/views/application_letters/_form.html.erb [hunk 1]
```diff
@@ -84,7 +91,7 @@
 
   <div class="form-group">
     <div class="col-lg-offset-2 col-lg-10">
-      <%= f.submit nil, :class => 'btn btn-primary' %>
+      <%= f.submit nil, :class => 'btn btn-primary', :disabled => @application_letter.after_deadline? %>
```
Comment (cmfcmf) [inline app/views/application_letters/_form.html.erb]: A user could simply remove the class and submit the form. A server-side check is needed!

**Review by cmfcmf — DISMISSED** [review]
LGTM, except for server-side checking.

File: app/models/application_letter.rb [hunk 2]
```diff
@@ -23,4 +23,12 @@ class ApplicationLetter < ActiveRecord::Base
 
   enum status: {accepted: 1, rejected: 0, pending: 2}
 
+  # Checks if the deadline is over
+  #
+  # @param none
+  # @return [Boolean] true if deadline is over
+  def after_deadline?
+    Date.today > event.application_deadline if event.present?
```
Comment (nstrelow) [inline app/models/application_letter.rb]: No, the same form is used for creating new applications which have no event

File: app/views/application_letters/_form.html.erb (hunk 1 above)
Comment (nstrelow) [inline app/views/application_letters/_form.html.erb]: Good point, will work on that

**Review by JuliusCosmoRomeo — APPROVED** [review]
seems to work properly (y)

File: app/models/application_letter.rb [hunk 3]
```diff
@@ -20,7 +20,23 @@ class ApplicationLetter < ActiveRecord::Base
   validates :grade, presence: true, numericality: { only_integer: true }
   validates :vegeterian, :vegan, :allergic, inclusion: { in: [true, false] }
   validates :vegeterian, :vegan, :allergic, exclusion: { in: [nil] }
+  validate :deadline_cannot_be_in_the_past, :if => Proc.new { |letter| !(letter.status_changed?) }
```
Comment (cmfcmf) [inline app/models/application_letter.rb]: What is `Proc` and why is it needed? 
It isn't clear to me why you used it (after reading [the docs](https://ruby-doc.org/core-2.2.0/Proc.html).

**Review by cmfcmf — COMMENTED** [review]
@nstrelow I would've liked to have a change to review the changes you made after I did my review. :)

**Comment by ekrebs5** [comment]
I noticed something while testing:
You are using a deadline attribute here. But at the moment there is no way to edit it in the UI, thus it is always nil, thus some of your functions fail. Because of this I cannot create an application letter.

In short, creating an application letter is not possible on dev right now. Could this be fixed somehow?
( @Baschdl )

**Comment by bjrne** [comment]
@ekrebs5 since this is already merged it could make sense to open an issue and link to this PR, this way the discussion of the problem probably moves to a more relevant place :) you could the either tag @Baschdl or all scrum masters if this depends on something another group built.

**Comment by cmfcmf** [comment]
> You are using a deadline attribute here. But at the moment there is no way to edit it in the UI, thus it is always nil, thus some of your functions fail. Because of this I cannot create an application letter.

This is [part of a different user story](https://github.com/hpi-swt2/workshop-portal/issues/18) which is not handled by our team. The @hpi-swt2/scrum-master-2016 should already know about the problem.

I agree, however, that it's difficult to use at the moment.

**Comment by nstrelow** [comment]
@ekrebs5 Thanks for noticing, I will check and fix issues I encounter.

**Comment by nstrelow** [comment]
See #170 and #171

File: app/models/application_letter.rb (hunk 3 above)
Comment (nstrelow) [inline app/models/application_letter.rb]: You need a Proc or lambda:
http://stackoverflow.com/questions/6232099/rails-validation-if-proc-new-or-lambda
http://stackoverflow.com/questions/8848601/rails-how-to-validate-a-field-only-if-a-another-field-has-a-certain-value

Comment (cmfcmf) [inline app/models/application_letter.rb]: I understand, thank you! I suppose a `lambda` would have been more "clear" for an inexperienced Rails developer like me. But anyways, good work!

**Pull request merged**